### PR TITLE
#394 Convert all test to JUnit5.

### PR DESCRIPTION
**Pull request by mvanaken**
To avoid confusion about which import to pick, decided to convert all tests to JUnit5. The steps consists of the following removals:

* Removal of the @Rule annotations and replaced it with assertThrows() test methods instead
* Removal of the @RunWith(Parameterized) for classes and replaced it with @ParameterizedTest on test methods.
* Changing imports from org.junit.(...) to org.juni.jupiter.api.Assertions.(...).

These steps required some additional changes to be made, especially around the [ParameterizedParse](https://github.com/parsingdata/metal/pull/395/files#diff-c7383c4315c5343a8d14883a0f87de903a6a11883d5d5c1b56b42b27348ac3a2) class.

Note: the old test suite before this branch contained 1147+16 tests and so does this run, so all tests are accounted for.

- [x] Find out why there are 27 more tests. - Checked the wrong commit. :see_no_evil: 
- [x] Change destination to master when #393 is merged.

### CODE REVIEW

File: core/src/test/java/io/parsingdata/metal/token/TokenRefTest.java
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/token/TokenRefTest.java:1]: Minor and unrelated: the method `createNamedTokens` can use some `final`s for the arguments.

Feel free to ignore this, I noticed that there are quite a few more places we could improve this.

File: core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java [hunk 1]
```diff
@@ -361,32 +355,36 @@ private static List<Object> instantiate(final List<Supplier<Object>> suppliers)
         return output;
     }
 
-    @Test
-    public void notEqualsNull() {
+    @ParameterizedTest
+    @MethodSource("data")
+    public void notEqualsNull(final Object object, final Object same, final Object[] other) {
         assertFalse(object.equals(null));
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java:361]: IntellIJ IDEA suggests `assertNotEquals(null, object);` here (and in a few other places in this test class, `ConstantSliceTest` and `EqualityTest`).

**Review by rdvdijk — APPROVED** [review]
Looks good!

File: core/src/test/java/io/parsingdata/metal/ImmutableListTest.java [hunk 2]
```diff
@@ -146,6 +146,7 @@ public void sizeEmpty() {
         assertEquals(0, new ImmutableList<ParseValue>().size);
     }
 
+
```
Comment (Carly-1) [inline core/src/test/java/io/parsingdata/metal/ImmutableListTest.java:149]: I think this change can be reverted.

**Review by Carly-1 — APPROVED** [review]
Nice improvement :+1:

File: core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java
[the remark this answers is not in the document]
Comment (mvanaken) [reply core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java:1]: I'll run a "Optimize imports" on entire project and then hoping my import settings for Metal or correct. :crossed_fingers: :grimacing:

File: formats/src/test/java/io/parsingdata/metal/format/FormatTest.java
[the remark this answers is not in the document]
Comment (mvanaken) [reply formats/src/test/java/io/parsingdata/metal/format/FormatTest.java:1]: Done!

File: core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java
[the remark this answers is not in the document]
Comment (mvanaken) [reply core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java:1]: Two files needed adjusting. Changed it! Tnx!

File: core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java (hunk 1 above)
Comment (mvanaken) [reply core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java:361]: The `equals` method is the method being tested here, so it is deliberate that the test is written like this. We want to ignore any other checks that is possibly checked when calling `assertNotEquals`. This is the test that really checks if the `equals` method returns a result as expected, so I am going to keep it as is.

File: core/src/test/java/io/parsingdata/metal/data/callback/CallbackTest.java
[the remark this answers is not in the document]
Comment (mvanaken) [reply core/src/test/java/io/parsingdata/metal/data/callback/CallbackTest.java:1]: Yes, there are many places where this is the case. I think I am going to leave that for a different issue. :)

File: core/src/test/java/io/parsingdata/metal/token/TokenRefTest.java
Comment (mvanaken) [reply core/src/test/java/io/parsingdata/metal/token/TokenRefTest.java:1]: Agreed! I'll ignore it for now. :wink: I'll see if there are many other places and if it is a lot, I'll create a separate issue for it.

**Pull request merged**